### ISSUE TITLE
Upcoming migration to LSP 1.2: drop-in replacement ?

### ISSUE THREAD

**Issue by v-fox** [sadko4u/lsp-plugins#168]
New LSP rewrite has quite a few separated libraries, as seen in https://github.com/sadko4u/lsp-plugins/issues/130#issuecomment-660188465 and https://github.com/sadko4u?tab=repositories but packaging all of them separately would be quite an annoyance for distro packagers. Could this be alleviated by some kind of collective repo, made-up of git submodules, that builds and installs all of them at once, as current version does ? It would be much easier to separate them into distro packages after collective install rather than to set-up separate builds.

**Comment by sadko4u** [comment]
Probably you're not familiar with the build process but it uses all repositories as submodules.
The only difference is, that necessary versions are taken from git by the following command:
```
make fetch
```

So distro builders are not required to build all libraries and after then build the plugins.
All what will be necessary is to issue the following set of commands:
```
make config # Configure build
make fetch # Fetch desired versions of source code from git
make # Build all necessary modules
make install # Install all produced binaries
```

Or, for full-tree tarball distribution we will be possible to omit step 2:
```
make config # Configure build
make # Build all necessary modules
make install # Install all produced binaries
```

This also makes possible to build plugins not as a complete bundle but also as separate bundle sets if plugins are split into separate git repositories as requested in #99.

**Comment by v-fox** [comment]
> Probably you're not familiar with the build process…

I'm surely not. I don't even know to which 1.2 repo this supposed to pertain.

> …but it uses all repositories as submodules.

If it takes something other than `git clone/checkout/pull` then it can't be used as submodules are intended to. There is also [a problem](https://gitlab.com/gitlab-org/gitlab/-/issues/540), which plagues some big projects with packaged shared-but-not-used-by-anything-else dependencies, of using mirrored repo that have links to other domain which is an issue of pulling github mirrors of gitlab-hosted things or vice versa but that's an issue of link formatting in repos.

That's… not great. OpenSUSE [Open Build Service](https://build.opensuse.org/package/show/home:X0F:branches:multimedia/lsp-plugins) uses only `git clone/checkout/pull` to make source archives for builds, it's was already inconvenient to build test version of LSP from temporary branches due to conflicts because of forced pushes and branch switches, that break `osc service disabledrun` command with which archives are usually made, but downloads inside source during build process are completely forbidden by OBS as they violate distro policy. That would require manually preparing archives by packaging user which itself violates assumption of bit-perfect repeatability of archives (not that it matters much by itself but having universal source-packaging way does).

This is one of the reasons why fancy SFLphone/Ring/Jami GNU SIP client [is not packaged and likely newer be](https://bugzilla.opensuse.org/show_bug.cgi?id=1174268). Distro maintainers do not need additional splits, package splits are supposed to be done in post-build scripts. And if you're building for something other then Gentoo, you would want all compile-time features all the time. #99, as [PW#420](https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/420) and such [proprietary users' desires](https://gist.github.com/abique/4c1b9b40f3413f0df1591d2a7c760db4) for sandboxing what supposed to be tied together but can't due to security concerns and misguided lust for… compactness (?) in their giant DAW-centric setups, are antithetical to purposes of mainline distros. For OBS, and other "normal" distro (Ubuntu, Fedora, etc.), purposes LSP are "split" enough as it is. Just as long we can keep generally unneeded VST and whatever parts away, it's "well split".

**Comment by sadko4u** [comment]
> That's… not great. OpenSUSE Open Build Service uses only git clone/checkout/pull to make source archives for builds

There's no problem of building **lsp-dsp-lib** which uses the same build process:
https://build.opensuse.org/package/show/home:ozu/lsp-dsp-lib

For LSP Plugins, **lsp-dsp-lib** will be built as a part of source code tree and won't require to build **lsp-dsp-lib** separately.

The problem with git modules is that they're hard to be versioned at all. But for my case, if you want to build some version, you just check out the corresponding tag, do ```make config && make fetch``` and all necessary versions of modules will be checked out into the tree.

**Comment by v-fox** [comment]
> make config && make fetch

As I've said, nothing more than automated git clone/checkout/pull is allowed in OBS's service commands. Sure you can do anything manually but it's a hassle for package maintenance. There is only few such packages now in the age of versioning systems that require this and they are major hassle for distributions. Especially if they get updated more often than once per year or need snapshots to be used.

> The problem with git modules is that they're hard to be versioned at all. 

Pretty sure that all projects that rely on submodules also set exact submodule versions that are tied to commits of the project. Not sure how exactly and how reliable that is but it works so far. Like https://github.com/RPCS3/rpcs3 and https://gitlab.linphone.org/BC/public/linphone-desktop

**Comment by sadko4u** [comment]
The LSP 1.2.0 release can be considered being a drop-in replacement for the previous 1.1.x branch.
There are still some bugs introduced in 1.2.0 but they will be fixed in 1.2.1.
